### PR TITLE
Add code coverage history support

### LINKED ISSUE #17086: Enable history for code coverage
We can optionally enable history for code coverage. See https://github.com/danielpalme/ReportGenerator/wiki/Integration#history. We might also instead use artifact upload/download - could be faster, but try both.

https://github.com/danielpalme/ReportGenerator/wiki/Integration#history

### PR DESCRIPTION
**Pull request by heaths**
Resolves #17086

### CODE REVIEW

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 1]
```diff
@@ -164,12 +168,25 @@ jobs:
...
+      - template: /eng/pipelines/templates/steps/select-feed.yml
+      - task: UniversalPackages@0
+        condition: and(succeededOrFailed(), eq(variables['CollectCoverage'], 'true'), eq(variables['coverage.collected'], 'true'))
+        displayName: Get code coverage history
+        continueOnError: true
+        inputs:
+          command: download
+          downloadDirectory: $(Coverage.HistoryDirectory)
+          feedsToUse: internal
+          vstsFeed: $DevOpsFeedID
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:180]:
```suggestion
          vstsFeed: $(DevOpsFeedID)
```
This is a devops variable so I think you need the parens.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 2]
```diff
@@ -179,3 +196,16 @@ jobs:
...
           summaryFileLocation: $(Build.ArtifactStagingDirectory)\coverage\Cobertura.xml
+      - ${{ if eq(variables['System.TeamProject'], 'internal') }}:
+        - task: UniversalPackages@0
+          condition: and(succeededOrFailed(), eq(variables['CollectCoverage'], 'true'), eq(variables['coverage.collected'], 'true'))
+          displayName: Publish code coverage history
+          inputs:
+            command: publish
+            publishDirectory: $(Coverage.HistoryDirectory)
+            feedsToUsePublish: internal
+            vstsFeedPublish: $DevOpsFeedID
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:207]:
```suggestion
            vstsFeedPublish: $(DevOpsFeedID)
```

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml (hunk 1 above)
Comment (heaths) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:180]: Oops

**Comment by pakrym** [comment]
Can you share some details about how this is expected to work? Where do I look for comparison?

**Comment by heaths** [comment]
@pakrym you won't find a comparison until we get one internal CI pushing code coverage history. And, honestly, I don't know what it will look like. I couldn't find any screenshots from various sites including the report generator's recommended author (which I mostly copied, but made it work for our monorepo/service directory pattern). I assume it'll look like deltas that test pass rates show, but lower in the code coverage section.

File: eng/pipelines/templates/steps/select-feed.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+parameters:
+  # Publish to https://dev.azure.com/azure-sdk/public/_packaging?_a=feed&feed=azure-sdk-for-net
+  DevOpsFeedId: '29ec6040-b234-4e31-b139-33dc4287b756/fa8c16a3-dbe0-4de2-a297-03065ec1ba3f'
+
+steps:
+- pwsh: |
+    # For safety default to publishing to the private feed.
+    # Publish to https://dev.azure.com/azure-sdk/internal/_packaging?_a=feed&feed=azure-sdk-for-net-pr
```
Comment (pakrym) [inline eng/pipelines/templates/steps/select-feed.yml:8]: Can we just use `Build.Repository.Name` in the URL?

Comment (pakrym) [reply eng/pipelines/templates/steps/select-feed.yml:8]: Oh, it's based on Id. got it.

File: eng/pipelines/templates/steps/select-feed.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+parameters:
+  # Publish to https://dev.azure.com/azure-sdk/public/_packaging?_a=feed&feed=azure-sdk-for-net
+  DevOpsFeedId: '29ec6040-b234-4e31-b139-33dc4287b756/fa8c16a3-dbe0-4de2-a297-03065ec1ba3f'
```
Comment (pakrym) [inline eng/pipelines/templates/steps/select-feed.yml:3]: Should we have this in some global variable file that we can override for the private repo?

Comment (heaths) [reply eng/pipelines/templates/steps/select-feed.yml:3]: I was considering it, but how it gets used here in a param is too soon to rely on global vars. Basically, this is the default. The only other place besides this file is where it's used in a pipeline that now includes this template, and I would've removed it except it's used unconditionally in one other place in that file.

That said, it's by far the only duplication we have. Unfortunate, but not uncommon.

File: eng/pipelines/templates/steps/select-feed.yml (hunk 3 above)
Comment (heaths) [reply eng/pipelines/templates/steps/select-feed.yml:8]: This is how AzDO does internal feeds - that's just the URL it resolves to. It has to be the tenant and feed ID.

**Comment by mitchdenny** [comment]
The long and the short of it is that the security model for Azure Pipelines/Azure DevOps means that you cannot publish to a pipeline that is running within the context of a forked PR. Because a forked PR can come from anywhere, we have to assume that they are potentially mallicious. And so Azure DevOps errs on the side of caution for forked PRs and instead of granting an access token which equals the same rights as the build service account, it create a token which expresses a severe subset of rights, just sufficient to post logs and artifacts associated with the build itself.

That said, the underlying storage technology for Universal Packages and Pipeline Artifacts is identical. The only difference between the two is that where a Universal Package has a name, a version and exists within a feed. A pipeline artifact has a name, and is associated with a pipeline run.

So why not use a pipeline artifact (we use them already in the build).

**Comment by heaths** [comment]
This is for a comparison, which means we have to publish only for the CI build. The PRs only need to download it. Using a build artifact means having some way to resolve to the last CI build. While I'm sure we can come up with something, it seems well beyond the passing interest in seeing historical data. @AlexGhiondea, thoughts?

**Comment by AlexGhiondea** [comment]
@heaths I agree -- I think the need for historical data is low. When we'll need specific questions answered from historical data we can revisit this.

**Pull request closed without merging**